client/optimizing.go

```diff
@@ -330,88 +331,200 @@ type watchResult struct {
+	active        []watchingClient
+	failed        []failedClient
+	retryInterval time.Duration
+}
+
+func (ws *watchState) loop(resultChan chan watchResult) {
+	closingClients := make(chan Client, 1)
+	defer close(resultChan)
+
+	// spin up initial watcher(s)
+	ws.tryRepopulate(resultChan, closingClients)
+
+	ticker := time.NewTicker(ws.optimizer.watchRetryInterval)
+	defer ticker.Stop()
 	for {
 		select {
-		case r, ok := <-in:
-			if !ok {
+		case c := <-closingClients:
+			// replace failed watchers
+			ws.done(c)
+			if ws.ctx.Err() == nil {
+				ws.tryRepopulate(resultChan, closingClients)
+			}
+			if len(ws.active) == 0 {
 				return
 			}
-
-			round := r.Result.Round()
-			timeOfRound := time.Unix(chain.TimeOfRound(info.Period, info.GenesisTime, round), 0)
-			stat := requestStat{
-				client:    r.Client,
-				rtt:       time.Since(timeOfRound),
-				startTime: timeOfRound,
+		case <-ticker.C:
+			// periodically cycle to fastest client.
+			if ws.has(ws.optimizer.fastestClients()[0]) == -1 {
+				ws.closeSlowest()
+				ws.tryRepopulate(resultChan, closingClients)
 			}
-			oc.updateStats([]*requestStat{&stat})
-			if round > latest {
-				latest = round
-				out <- r.Result
+		case <-ws.ctx.Done():
+			for _, c := range ws.active {
+				c.CancelFunc()
 			}
-		case <-done:
+			ws.active = []watchingClient{}
+		}
+	}
+}
+
+func (ws *watchState) tryRepopulate(results chan watchResult, done chan Client) {
+	ws.clean()
+
+	for {
+		if len(ws.active) >= ws.optimizer.requestConcurrency {
 			return
 		}
+		c := ws.nextUnwatched()
+		if c == nil {
+			return
+		}
+		cctx, cancel := context.WithCancel(ws.ctx)
+
+		ws.active = append(ws.active, watchingClient{c, cancel})
+		ws.optimizer.log.Info("optimizing_client", "watching on client", "client", fmt.Sprintf("%s", c))
+		go ws.watchNext(cctx, c, results, done)
 	}
 }
 
-// Watch returns new randomness as it becomes available.
-func (oc *optimizingClient) Watch(ctx context.Context) <-chan Result {
-	outChan := make(chan Result)
-	inChan := make(chan watchResult)
-	wg := sync.WaitGroup{}
+func (ws *watchState) watchNext(ctx context.Context, c Client, out chan watchResult, done chan Client) {
+	defer func() { done <- c }()
 
-	info, err := oc.Info(ctx)
-	if err != nil {
-		oc.log.Error("optimizing_client", "failed to learn info", "err", err)
-		close(outChan)
-		return outChan
+	resultStream := c.Watch(ctx)
+	for r := range resultStream {
+		out <- watchResult{r, c}
 	}
+	ws.optimizer.log.Info("optimizing_client", "watch ended", "client", fmt.Sprintf("%s", c))
+}
 
-	for _, c := range oc.clients {
-		wg.Add(1)
-		go oc.watchIngest(ctx, c, inChan, &wg)
+func (ws *watchState) clean() {
+	nf := make([]failedClient, 0, len(ws.failed))
+	for _, f := range ws.failed {
+		if f.Time.After(time.Now()) {
+			nf = append(nf, f)
+		}
 	}
+	ws.failed = nf
+}
 
-	doneChan := make(chan bool)
-	go func() {
-		wg.Wait()
-		close(doneChan)
-	}()
+func (ws *watchState) close(clientIdx int) {
```

Comment:
Take a `Client` instead of an index and make `ws.has(Client)` return a bool?

Reply:
i think that would end up needing to scan through the list of clients twice. seems nicer to only do that work once